patient-appointment-management/functions/helpers.private.js

```diff
@@ -0,0 +1,312 @@
+  _GLUE_CRAWLER_BASE: `twilio-${THIS_APPLICATION_NAME}`,
+  _GLUE_DATABASE_BASE: 'patient_appointments',
+};
+
+function determineRuntimeEnvironment(context) {
+  if (context === null || Object.keys(context).length === 0) {
+    return 'local';
+  } else if (
+    context.DOMAIN_NAME &&
+    context.DOMAIN_NAME.startsWith('localhost')
+  ) {
+    return 'localhost';
+  }
+  return 'server';
+}
+
+/*
+ * --------------------------------------------------------------------------------
+ *
+ * --------------------------------------------------------------------------------
+ */
+async function setParam(context, key, value) {
+  const onLocalhost = Boolean(
+    context.DOMAIN_NAME && context.DOMAIN_NAME.startsWith('localhost')
+  );
+  console.debug('Runtime environment is localhost:', onLocalhost);
+
+  switch (determineRuntimeEnvironment(context)) {
+    case 'local':
+      process.env[key] = value;
+      break;
+    case 'localhost':
+      process.env[key] = value;
+      break;
+    case 'server':
+      const client = context.getTwilioClient();
+      console.log('twilioClient', client);
+      // eslint-disable-next-line no-use-before-define
+      const service_sid = await getParam(context, 'TWILIO_SERVICE_SID');
+      console.log('service_sid', service_sid);
+      // eslint-disable-next-line no-use-before-define
+      const environment_sid = await getParam(context, 'TWILIO_ENVIRONMENT_SID');
+      console.log('environment_sid', environment_sid);
+
+      let variable_sid = null;
+      await client.serverless
+        .services(service_sid)
+        .environments(environment_sid)
+        .variables.list()
+        .then((variables) =>
+          variables.forEach((v) => {
+            if (v.key === key) variable_sid = v.sid;
+          })
+        );
+
+      if (variable_sid === null) {
+        await client.serverless
+          .services(service_sid)
+          .environments(environment_sid)
+          .variables.create({ key, value })
+          .then((v) => console.log('Created variable', v.key));
+      } else {
+        await client.serverless
+          .services(service_sid)
+          .environments(environment_sid)
+          .variables(variable_sid)
+          .update({ value })
+          .then((v) => console.log('Updated variable', v.key));
+      }
+      break;
+    default:
+      break;
+  }
+  console.log('Assigned', key, '=', value);
+}
+
+/*
+ * --------------------------------------------------------------------------------
+ * retrieve environment variable from
+ * - service environment (aka context)
+ * - os for local development (via os.process)
+ * - per resource-specific logic
+ *   . TWILIO_SERVICE_SID from context, otherwise matching application name
+ *   . TWILIO_ENVIRONMENT_SID from TWILIO_SERVICE_SID
+ *   . TWILIO_ENVIRONMENT_DOMAIN_NAME from TWILIO_SERVICE_SID
+ * --------------------------------------------------------------------------------
+ */
+async function getParam(context, key) {
```

Comment:
I don't think I understand the logic here. Is this entire logic so that you can get a value that has been updated during the same execution? If so why not just update `context.` in the `setParam` and fetch it from memory? This isn't very scalable since you keep hitting the Twilio API.

Reply:
You are correct, when I was developing updating the `context.` didn't quite work. The services seem to behavior differently when running via twilio cli vs after it is deployed ... will need to figure this one out too.